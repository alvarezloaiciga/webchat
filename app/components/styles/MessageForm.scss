.MessageForm {
  width: 100%;
  flex: 0 0 auto;
  background: #fff;
  display: flex;
  flex-direction: column;

  .messageArea {
    display: flex;
    flex: 0 0 auto;

    textarea {
      padding: 12px 0 12px 12px;
      flex: 1 1 auto;
      width: 100%;
      border: none;
      resize: none;
      font-size: 14px;

      &:focus {
        outline: none;
      }
    }

    .sendBtn {
      padding: 12px;
      flex: 0 0 auto;
      border: none;
      font-size: 14px;
      background: #fff;
      font-weight: bold;
      transition: 0.15s ease-in-out color;

      &:not([disabled]) {
        color: #59ad5d;
      }
    }
  }

  .poke {
    margin-top: auto;
    background: white;
    font-size: 14px;
    border-bottom: 1px solid #eee;
    animation: fadeUp 1.0s 1;

    .pokeBody {
      padding: 7px 9px;

      .TypingIndicator {
        margin-left: 5px;
        vertical-align: text-bottom;
      }
    }
  }
}

// Total Width: 400px
// Total Height: 600px
// MessageForm: 120px
// Banner: 40px
// Transcript: 440px
.noFlexbox .MessageForm {
  height: 120px;
  min-height: 120px;
  max-height: 120px;

  .poke {
    min-height: 30px;
    max-height: 30px;
    height: 30px;
    display: block;
  }

  .messageArea {
    display: table;

    textarea {
      max-height: 60px;
      min-height: 60px;
      height: 60px;
      overflow: auto;
      width: 318px; // 318 + 12padding = 330
    }

    .sendBtn {
      display: table-cell;
      vertical-align: middle;
      width: 46px; // 46 + 24padding = 70
    }
  }
}
