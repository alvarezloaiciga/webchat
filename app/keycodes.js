// @flow

type keycodesType = {
  backspace: number,
  tab: number,
  enter: number,
  shift: number,
  ctrl: number,
  alt: number,
  pausebreak: number,
  capslock: number,
  esc: number,
  space: number,
  pageup: number,
  pagedown: number,
  end: number,
  home: number,
  leftarrow: number,
  uparrow: number,
  rightarrow: number,
  downarrow: number,
  insert: number,
  delete: number,
  '0': number,
  '1': number,
  '2': number,
  '3': number,
  '4': number,
  '5': number,
  '6': number,
  '7': number,
  '8': number,
  '9': number,
  a: number,
  b: number,
  c: number,
  d: number,
  e: number,
  f: number,
  g: number,
  h: number,
  i: number,
  j: number,
  k: number,
  l: number,
  m: number,
  n: number,
  o: number,
  p: number,
  q: number,
  r: number,
  s: number,
  t: number,
  u: number,
  v: number,
  w: number,
  x: number,
  y: number,
  z: number,
  leftwindowkey: number,
  rightwindowkey: number,
  selectkey: number,
  numpad0: number,
  numpad1: number,
  numpad2: number,
  numpad3: number,
  numpad4: number,
  numpad5: number,
  numpad6: number,
  numpad7: number,
  numpad8: number,
  numpad9: number,
  multiply: number,
  add: number,
  subtract: number,
  decimalpoint: number,
  divide: number,
  f1: number,
  f2: number,
  f3: number,
  f4: number,
  f5: number,
  f6: number,
  f7: number,
  f8: number,
  f9: number,
  f10: number,
  f11: number,
  f12: number,
  numlock: number,
  scrolllock: number,
  semicolon: number,
  equalsign: number,
  comma: number,
  dash: number,
  period: number,
  forwardslash: number,
  graveaccent: number,
  openbracket: number,
  backslash: number,
  closebracket: number,
  singlequote: number,
};

const keycodes: keycodesType = {
  backspace: 8,
  tab: 9,
  enter: 13,
  shift: 16,
  ctrl: 17,
  alt: 18,
  pausebreak: 19,
  capslock: 20,
  esc: 27,
  space: 32,
  pageup: 33,
  pagedown: 34,
  end: 35,
  home: 36,
  leftarrow: 37,
  uparrow: 38,
  rightarrow: 39,
  downarrow: 40,
  insert: 45,
  delete: 46,
  '0': 48, // eslint-disable-line quote-props
  '1': 49, // eslint-disable-line quote-props
  '2': 50, // eslint-disable-line quote-props
  '3': 51, // eslint-disable-line quote-props
  '4': 52, // eslint-disable-line quote-props
  '5': 53, // eslint-disable-line quote-props
  '6': 54, // eslint-disable-line quote-props
  '7': 55, // eslint-disable-line quote-props
  '8': 56, // eslint-disable-line quote-props
  '9': 57, // eslint-disable-line quote-props
  a: 65,
  b: 66,
  c: 67,
  d: 68,
  e: 69,
  f: 70,
  g: 71,
  h: 72,
  i: 73,
  j: 74,
  k: 75,
  l: 76,
  m: 77,
  n: 78,
  o: 79,
  p: 80,
  q: 81,
  r: 82,
  s: 83,
  t: 84,
  u: 85,
  v: 86,
  w: 87,
  x: 88,
  y: 89,
  z: 90,
  leftwindowkey: 91,
  rightwindowkey: 92,
  selectkey: 93,
  numpad0: 96,
  numpad1: 97,
  numpad2: 98,
  numpad3: 99,
  numpad4: 100,
  numpad5: 101,
  numpad6: 102,
  numpad7: 103,
  numpad8: 104,
  numpad9: 105,
  multiply: 106,
  add: 107,
  subtract: 109,
  decimalpoint: 110,
  divide: 111,
  f1: 112,
  f2: 113,
  f3: 114,
  f4: 115,
  f5: 116,
  f6: 117,
  f7: 118,
  f8: 119,
  f9: 120,
  f10: 121,
  f11: 122,
  f12: 123,
  numlock: 144,
  scrolllock: 145,
  semicolon: 186,
  equalsign: 187,
  comma: 188,
  dash: 189,
  period: 190,
  forwardslash: 191,
  graveaccent: 192,
  openbracket: 219,
  backslash: 220,
  closebracket: 221,
  singlequote: 222,
};

export default keycodes;
