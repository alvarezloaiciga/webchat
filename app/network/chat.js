// @flow

import { QUIQ } from 'utils/utils';
import fetch from 'isomorphic-fetch';
import type { Conversation } from 'types';

const { TENANT, HOST, CONTACT_POINT } = QUIQ;

export const joinChat = () => {
  fetch(`https://${TENANT}.${HOST}/api/v1/messaging/chat/${CONTACT_POINT}/join`, {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  });
};

export const leaveChat = () => {
  fetch(`https://${TENANT}.${HOST}/api/v1/messaging/chat/${CONTACT_POINT}/leave`, {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  });
};

export const poll = (callback: (conversation: Conversation) => void) => {
  fetch(`https://${TENANT}.${HOST}/api/v1/messaging/chat/${CONTACT_POINT}`, {
    mode: 'cors',
    credentials: 'include',
  }).then(response => response.json().then(callback));
};

export const ping = () => {
  fetch(`https://${TENANT}.${HOST}/api/v1/webchat/endpoints/${CONTACT_POINT}/socket-ping`, {
    mode: 'cors',
    method: 'POST',
    credentials: 'include',
  });
};

export const addMessage = (body: string) => {
  fetch(`https://${TENANT}.${HOST}/api/v1/messaging/chat/${CONTACT_POINT}/send-message`, {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({body}),
  });
};

export const retrieveMessages = (
  callback: (conversation: Conversation, resolve: () => void) => void
) => new Promise((resolve, reject) => {
  console.log('Initial message retrieval starting'); // eslint-disable-line no-console
  fetch(`https://${TENANT}.${HOST}/api/v1/messaging/chat/${CONTACT_POINT}`, {
    mode: 'cors',
    credentials: 'include',
  }).then(response => {
    response.json().then(res => {
      if (res.status && res.status >= 300) {
        return reject();
      }

      callback(res, resolve);
    });
  }).catch(() => {
    reject();
  });
});
