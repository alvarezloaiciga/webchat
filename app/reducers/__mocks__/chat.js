import {getMockConfiguration} from 'utils/testHelpers';

export const getConfiguration = () => getMockConfiguration();
export const getMessage = message => getMockConfiguration().messages[message].defaultMessage;
