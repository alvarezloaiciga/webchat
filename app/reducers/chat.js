// @flow
import {inStandaloneMode} from 'Common/Utils';
import {createStore} from 'redux';
import {ChatInitializedState} from 'Common/Constants';
import quiqOptions from 'utils/quiq';
import type {ChatState, Action, ChatInitializedStateType, Message} from 'types';

type ChatAction = {
  chatContainerHidden?: boolean,
  agentsAvailable?: boolean,
  initializedState?: ChatInitializedStateType,
  transcript?: Array<Message>,
  agentTyping?: boolean,
};

export const initialState = {
  chatContainerHidden: true,
  agentsAvailable: false,
  initializedState: ChatInitializedState.UNINITIALIZED,
  transcript: [],
  agentTyping: false,
  welcomeFormRegistered: !quiqOptions.welcomeForm,
};

const chat = (state: ChatState, action: Action & ChatAction) => {
  switch (action.type) {
    case 'CHAT_CONTAINER_HIDDEN':
      return Object.assign({}, state, {
        chatContainerHidden: inStandaloneMode() ? false : action.chatContainerHidden,
      });
    case 'AGENTS_AVAILABLE':
      return Object.assign({}, state, {
        agentsAvailable: action.agentsAvailable,
      });
    case 'CHAT_INITIALIZED_STATE': {
      if (state.initializedState === ChatInitializedState.BURNED) {
        // One does not simply become unburned.
        return state;
      }

      return Object.assign({}, state, {initializedState: action.initializedState});
    }
    case 'UPDATE_TRANSCRIPT':
      return Object.assign({}, state, {transcript: action.transcript || []});
    case 'AGENT_TYPING':
      return Object.assign({}, state, {agentTyping: action.agentTyping});
    case 'WELCOME_FORM_REGISTERED':
      return Object.assign({}, state, {welcomeFormRegistered: true});
    case 'NEW_WEBCHAT_SESSION':
      // Reset state to initial state.
      // We keep the visibility state from before the new session, and we set initialized state to LOADING (while socket reconnects)
      return Object.assign(
        {},
        {...initialState},
        {
          chatContainerHidden: state.chatContainerHidden,
          agentsAvailable: state.agentsAvailable,
          initializedState: ChatInitializedState.LOADING,
        },
      );
    default:
      return state;
  }
};

export default chat;

// Selectors
export const getChatContainerHidden = (state: ChatState): boolean => {
  return state.chatContainerHidden;
};

export const getAgentsAvailable = (state: ChatState): boolean => {
  return state.agentsAvailable;
};