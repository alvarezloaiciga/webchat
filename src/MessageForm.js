import React, {Component} from 'react';
import Textarea from 'react-textarea-autosize';

export class MessageForm extends Component {
  state = {
    text: '',
  };

  handleTextChanged = (e) => {
    this.setState({text: e.target.value});
  }

  render() {
    return (
      <div className="messageForm">
        <Textarea
          name="message"
          value={this.state.text}
          onChange={this.handleTextChanged}
          placeholder="Send us a message..."
        />
      </div>
    );
  }
}

export default MessageForm;
