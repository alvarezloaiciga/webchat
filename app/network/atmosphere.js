// @flow
/* eslint-disable no-console */

import atmosphere from 'atmosphere.js';
import { ping } from 'network/chat';
import type { AtmosphereRequest, AtmosphereConnection, AtmosphereConnectionBuilder, AtmosphereMessage } from 'types';

let connection: AtmosphereConnection;
let onConnectionLoss: () => void;
let onConnectionEstablish: () => void;
let handleMessage: (message: AtmosphereMessage) => void;

const buildRequest = (socketUrl: string) => ({
  url: `https://${socketUrl}`,
  enableXDR: true,
  withCredentials: true,
  contentType: 'application/json',
  logLevel: 'error',
  transport: 'websocket',
  fallbackTransport: 'long-polling',
  trackMessageLength: true,
  maxReconnectOnClose: Number.MAX_SAFE_INTEGER,
  // Keep reconnectInterval at 10 seconds.  Otherwise if API-GW drops the atmosphere connection,
  // we will hammer them with onReconnect requests. See SER-2620
  reconnectInterval: 10000,
  onOpen, onClose, onReopen, onReconnect, // eslint-disable-line no-use-before-define
  onMessage, onTransportFailure, onError, onClientTimeout, // eslint-disable-line no-use-before-define
});


export const connectSocket = (builder: AtmosphereConnectionBuilder) => {
  const { socketUrl, options } = builder;

  onConnectionLoss = options.onConnectionLoss;
  onConnectionEstablish = options.onConnectionEstablish;
  handleMessage = options.handleMessage;

  connection = { ...atmosphere.subscribe(buildRequest(socketUrl)) };
};

export const disconnectSocket = () => {
  atmosphere.unsubscribe();
};

// ******** Atmosphere callbacks *********
const onOpen = (response) => {
  // Carry the UUID. This is required if you want to call subscribe(request) again.
  connection.request.uuid = response.request.uuid;
  onConnectionEstablish && onConnectionEstablish();
  console.log('Socket open');
};

const onReopen = () => {
  console.log('Socket reopened');
  onConnectionEstablish && onConnectionEstablish();
};

const onReconnect = (req: AtmosphereRequest) => {
  // Long-polling doesn't clear up the error until it gets something back from the server
  // Force this to happen by sending a ping.
  clearTimeout(connection.pingTimeout);
  if (req.transport === 'long-polling') {
    connection.pingTimeout = setTimeout(() => {
      ping();
      console.log('Atmosphere: ping sent');
    }, connection.request.reconnectInterval + 5000);
  }
  console.log('Atmosphere: reconnect');
};

const onMessage = (res) => {
  let message;
  try {
    message = atmosphere.util.parseJSON(res.responseBody);
    handleMessage && handleMessage(message);
  } catch (e) {
    console.error('Error parsing Quiq websocket message');
    return;
  }

  console.log(message);
};

const onTransportFailure = (errorMsg: string, req: AtmosphereRequest) => {
  console.log(`Transport failed: ${req.transport}`);
};

const onError = () => {
  console.log('Atmosphere error');
  onConnectionLoss && onConnectionLoss();
};

const onClientTimeout = () => {
  console.log('Atmosphere client timeout');
  onConnectionLoss && onConnectionLoss();
};

const onClose = () => {
  console.log('Atmosphere connection closed');
};
