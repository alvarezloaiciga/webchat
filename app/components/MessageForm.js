// @flow

import React, {Component} from 'react';
import { QUIQ } from 'utils/utils';
import { addMessage } from 'network/chat';
import keycodes from 'keycodes';
import Textarea from 'react-textarea-autosize';
import messages from 'messages';
import { formatMessage } from 'core-ui/services/i18nService';
import { FormattedMessage } from 'react-intl';
import './styles/MessageForm.scss';

const { COLOR } = QUIQ;
type MessageFormState = {
  text: string,
};

export class MessageForm extends Component {
  state: MessageFormState = {
    text: '',
  };

  handleTextChanged = (e: SyntheticInputEvent) => {
    this.setState({text: e.target.value});
  }

  addMessage = () => {
    const text = this.state.text.trim();
    if (text) {
      addMessage(text);
      this.setState({text: ''});
    }
  }

  handleKeyDown = (e: SyntheticKeyboardEvent) => {
    if (e.keyCode === keycodes.enter) {
      e.preventDefault();
      this.addMessage();
    }
  }

  render() {
    return (
      <div className="MessageForm">
        <Textarea
          name="message"
          value={this.state.text}
          onChange={this.handleTextChanged}
          onKeyDown={this.handleKeyDown}
          placeholder={ formatMessage(messages.sendUsAMessage) }
        />
        <button
          className="sendBtn"
          onClick={this.addMessage}
          disabled={this.state.text.trim() === ''}
          style={{ color: COLOR }}
        >
          <FormattedMessage { ...messages.send } />
        </button>
      </div>
    );
  }
}

export default MessageForm;
