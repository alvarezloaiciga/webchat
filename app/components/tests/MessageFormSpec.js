// @flow

import React from 'react';
import keycodes from 'keycodes';
import MessageForm from '../MessageForm';
import { addMessage } from 'network/chat';
import { shallow } from 'enzyme';
import type { ShallowWrapper } from 'enzyme';

jest.mock('network/chat');

describe('MessageForm component', () => {
  let wrapper:ShallowWrapper;
  let instance: any;
  let render: () => void;

  beforeEach(() => {
    render = () => {
      wrapper = shallow(<MessageForm />);
      instance = wrapper.instance();
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('adding text', () => {
    it('adds text', () => {
      wrapper.setState({ text: 'Tool Time' });
      expect(wrapper.find('TextareaAutosize').prop('value')).toBe('Tool Time');
    });

    describe('pressing enter', () => {
      beforeEach(() => {
        wrapper.setState({ text: 'OOHOOOHOOH' });
        instance.handleKeyDown({ keyCode: keycodes.enter, preventDefault: jest.fn() });
      });

      it('adds text', () => {
        expect(addMessage).toBeCalledWith('OOHOOOHOOH');
      });

      it('clears message form', () => {
        wrapper.update();
        expect(wrapper.find('TextareaAutosize').prop('value')).toBe('');
      });
    });
  });
});
