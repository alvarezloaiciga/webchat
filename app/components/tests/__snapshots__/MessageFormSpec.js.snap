// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm component rendering renders 1`] = `
<div
  className="MessageForm"
>
  <TextareaAutosize
    name="message"
    onChange={[Function]}
    onHeightChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Send us a message..."
    useCacheForDOMMeasurements={false}
    value=""
  />
  <button
    className="sendBtn"
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "color": "#333",
      }
    }
  >
    <FormattedMessage
      defaultMessage="Send"
      description="Generic send message"
      id="send"
      values={Object {}}
    />
  </button>
</div>
`;
